stdlib/asm/mem.masm

```diff
@@ -49,3 +49,132 @@ export.memcopy
+  end
+
+  movup.13 drop # (5 cycles)
+  # [C', B', A', write_ptr', ...]
+end
+
+#! Copies an arbitrary number of words from the advice tape to memory
+#!
+#! Input: [num_words, write_ptr, ...]
+#! Output: [HASH, write_ptr', ...]
+#! Cycles:
+#!  even num_words: 41 + 8 * num_words
+#!  odd num_words: 58 + 8 * num_words
+export.move_words_from_advice_stack_to_memory.0
+  # check if there is an odd number of words (6 cycles)
+  dup is_odd
+  # => [is_odd, num_words, write_ptr, ...]
+
+  # copy is_odd, it defines if last last word requires padding (2 cycles)
+  dup movdn.3
+  # => [is_odd, num_words, write_ptr, needs_padding, ...]
+
+  # make `num_words` is even (2 cycles)
+  sub swap
+  # => [words_even, write_ptr, needs_padding, ...]
+
+  # Prepare the capacity word. For rescue prime optimized the first element is
+  # set to `1` when padding is used and `0` otherwse, this is determined by the
+  # `needs_padding` flag, so just copy it. (4 cycles)
+  dup.2 push.0.0.0
+  # => [A, write_ptr, words_even, needs_padding, ...]
+
+  # set initial hasher state (8 cycles)
+  padw padw
+  # => [C, B, A, write_ptr, words_even, needs_padding, ...]
+
+  # (8 + 11 * num_words cycles)
+  exec.move_even_words_from_advice_stack_to_memory
+  # => [C, B, A, write_ptr, needs_padding, ...]
+
+  # (4 cycles)
+  movup.13
+  # => [needs_padding, C, B, A, write_ptr, ...]
+
+  # if(needs_padding) (17 cycles)
+  if.true
+    # Rescue Prime Optimized uses overwrite mode, drop `C`. (4 cycles)
+    dropw
+    # => [B, A, write_ptr, ...]
+
+    # Overwrite the `B` with the new data (1 cycles)
+    adv_loadw
+    # => [B', A, write_ptr, ...]
+
+    # - get the memory address that B' should be saved to
+    # - update the write_ptr to point to the next address (4 cycles)
+    movup.8 dup.0 add.1 movdn.5
+    # => [write_ptr, B', write_ptr+1, A, ...]
+
+    # save data to memory (1 cycles)
+    mem_storew
+    # => [B', write_ptr+1, A, ...]
+
+    # Fix write_ptr position (2 cycles)
+    movup.4 movdn.8
+    # => [B', A, write_ptr+1, ...]
+
+    # Push padding word (4 cycles)
+    push.1.0.0.0
+    # => [C, B', A, write_ptr+1, ...]
+
+    # Run RPO permutation (1 cycles)
+    hperm
+    # => [C', B', A', write_ptr+1, ...]
+  end
+
+  # The RPO result is word B, discard the unused portion of the rate and the capacity. (9 cycles)
+  dropw swapw dropw
+  # => [rpo_hash, write_ptr', ...]
+end
+
+#! Moves an arbitrary number of words from the advice tape to memory and asserts it matches the commitment.
+#!
+#! Input: [num_words, write_ptr, commitment, ...]
+#! Output: [write_ptr', ...]
+#! Cycles:
+#!  even num_words: 53 + 8 * num_words
+#!  odd num_words: 70 + 8 * num_words
+export.move_words_from_advice_stack_to_memory_with_commitment.0
+  # Copies the tape data to memory
+  # stack: [rpo_hash, write_ptr', commitment, ...]
+  exec.move_words_from_advice_stack_to_memory
+
+  # Save the write_ptr
+  # stack: [rpo_hash, commitment, write_ptr', ...] (2 cycles)
+  movup.4 movdn.8
```

Comment:
The commenting convention here (and in other places in this file) differs from what we are trying to follow. For example, I think this should have looked more like this:
```
# Save the write_ptr (2 cycles)
movup.4 movdn.8
# => [rpo_hash, commitment, write_ptr', ...]
```